File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    public String[] solution(String[] orders, int[] course) {
        
        ArrayList<String> result = new ArrayList<>();

        for (int size : course) {

            HashMap<String, Integer> combsCount = new HashMap<>();

            for (String order : orders) {
                char[] items = order.toCharArray();
                Arrays.sort(items);
                solve(items, size, combsCount);
            }

            int max = 0;
            for (int count : combsCount.values()) {
                if (count >= 2 && count > max) {
                    max = count;
                }
            }

            for (String key : combsCount.keySet()) {
                if (combsCount.get(key) == max && max >= 2) {
                    result.add(key);
                }
            }
        }

        Collections.sort(result);

        String[] ans = new String[result.size()];
        for (int i = 0; i < result.size(); i++) {
            ans[i] = result.get(i);
        }

        return ans;
    }
    
    public void solve(char[] items, int size, HashMap<String, Integer> combsCount) {

        ArrayList<String> combs = new ArrayList<>();

        getCombs(items, size,0,"", combs);

        for (String comb : combs) {
            combsCount.put(comb, combsCount.getOrDefault(comb, 0) + 1);
        }
    }
    
    public void getCombs(char[] items, int size, int start, String now,  ArrayList<String> combs) {
        if (size == 0) {
            combs.add(now);
            return;
        }

        for (int i = start; i < items.length; i++) {
            getCombs(items, size - 1, i + 1, now + items[i], combs);
        }
    }
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from collections import defaultdict

def solution(orders, course):
    
    result = []
    
    for size in course:
        
        combs_count = defaultdict(int)
        
        for order in orders:
            items = list(order)
            
            if len(items) < size:
                continue
            
            items.sort()
            solve(items, size, combs_count)
        
        max_count = 0
        for count in combs_count.values():
            if count >= 2 and count > max_count:
                max_count = count
        
        for comb, count in combs_count.items():
            if count == max_count:
                result.append(comb)
    
    result.sort()
    return result

def solve(items, size, combs_count):
    combs = []
    get_combs(items, size, 0, "", combs)

    for comb in combs:
        combs_count[comb] += 1

def get_combs(items, size, start, now, combs):
    if size == 0:
        combs.append(now)
        return

    for i in range(start, len(items)):
        get_combs(items, size - 1, i + 1, now + items[i], combs)